src/plots/Volume/avtVolumeRenderer.C

```diff
@@ -135,374 +96,121 @@ avtVolumeRenderer::New(void)
-//    Account for Log scaling when determining vmin and vmax. 
-//
-//    Hank Childs, Tue Feb  6 15:39:01 PST 2007
-//    Treat the values for "min" and "max" as min and max, not as log(min),
-//    log(max).  That is 6 was being treated as 10^6.  No longer.
-//
-//    Brad Whitlock, Mon Dec 15 13:24:12 PST 2008
-//    I moved most of the code to helper functions.
-//
-//    Jeremy Meredith, Tue Jan  5 15:48:41 EST 2010
-//    Output un-normalized gradient magnitude and actual calculated maximum.
-//
-//    Allen Harvey, Thurs Nov 3 7:21:13 EST 2011
-//    Added support for holding a compact support variable
-//
-//    Alister Maguire, Thu Sep 14 13:36:16 PDT 2017
-//    Added a check for 2D data.
-//
-// ****************************************************************************
-
-void
-avtVolumeRenderer::Initialize(vtkDataSet *ds)
-{
-    StackTimer t("avtVolumeRenderer::Initialize");
-
-    vtkDataArray *data = 0, *opac = 0;
-    if(!VolumeGetScalars(atts, ds, data, opac))
-        return;
+        if( dataArr == nullptr )
+        {
+            EXCEPTION1(InvalidVariableException, "");
+        }
+    }
 
-    VolumeGetVariableExtents(atts, data,
-        this->varmin, this->varmax, 
-        this->vmin, this->vmax, this->vsize);
+    // There could be a spearate opacity scalar data array.
+    std::string activeVariable  = dataArr->GetName();
+    std::string opacityVariable = m_atts.GetOpacityVariable();
 
-    // Get the opacity variable's extents.
-    VolumeGetOpacityExtents(atts, opac,
-        this->omin, this->omax, this->osize);
+    // Before calling NeedImage the number of components needs to be
+    // known which is done by comparing the data and opacity variable
+    // names.
+    NumberOfComponents(activeVariable, opacityVariable);
 
-    // calculate gradient
-    if (ds->GetDataObjectType() == VTK_RECTILINEAR_GRID)
+    // The data and opacity ranges must be known before calling
+    // UpdateRenderingState.
+    if( NeedImage(in_ds) )
     {
-        vtkRectilinearGrid *grid = (vtkRectilinearGrid*)ds;
-        int dims[3];
-        grid->GetDimensions(dims);
+        // Get the local data range so to ignore NO_DATA_VALUE values.
+        // dataArr->GetRange( m_dataRange );
+        int nTupples = dataArr->GetNumberOfTuples();
 
-        if (atts.GetLightingFlag() && gm == NULL) // make sure the gradient was invalidated first
-        {
+        m_dataRange[0] = +FLT_MAX;
+        m_dataRange[1] = -FLT_MAX;
 
-            int nels = dims[0] * dims[1] * dims[2];
-            gx  = new float[nels];
-            gy  = new float[nels];
-            gz  = new float[nels];
-            gm  = new float[nels];
-            gmn = new float[nels];
-            hs = NULL;
-            float ghostval = omax+osize;
-            gm_max = VolumeCalculateGradient(atts, grid, opac, gx, gy, gz, gm, gmn, ghostval);
+        for( int ptId = 0; ptId<nTupples; ++ptId )
+        {
+            double dataTuple = dataArr->GetTuple1(ptId);
```

Comment:
`NO_DATA_VALUE`'s value is close to the largest negative 32 bit float. But, it is way far from the largest negative 64 bit double. So, this means that a large fraction of the total range of values representable by a double is considered out of bounds. Is this correct?